pesudocode/voxel_grid_structure_3.hpp

```diff
@@ -0,0 +1,130 @@
+#include <iostream>
+#include <type_traits>
+#include <vector>
+#include <unordered_map>
+#include <Eigen/Dense>
+
+class PointXYZ
+{
+public:
+    PointXYZ() {}
+};
+
+template <typename PointT>
+class PointCloudT
+{
+public:
+    PointCloudT() {}
+
+    void push_back(PointT p) {}
+};
+
+template <typename PointT>
+class Filter
+{
+    using PointCloud = PointCloudT<PointXYZ>;
+
+public:
+    Filter() {}
+    virtual void applyFilter(PointCloud &pc) = 0;
+    void filter(PointCloud &output) { applyFilter(output); }
+};
+
+// ================== Base of VoxelGrid ==================
+
+// Check point type by looking at typedef/using in the GridStruct,
+// require user to define it
+#define GET_POINT_TYPE(GridStruct) \
+    typename GridStruct::PointType
+
+template <typename GridStruct, typename Grids, typename PointT = GET_POINT_TYPE(GridStruct)>
+class GridFilter : public GridStruct, public Filter<PointT>
+{
+    // TODO:
+    // check if the required functions exist in Grid
+    // static_assert(...);
+
+    using PointCloud = PointCloudT<PointT>;
+
+public:
+    GridFilter() : GridStruct(input_, grids_)
+    {
+    }
+
+protected:
+    void applyFilter(PointCloud &output)
+    {
+        // move setup to the contructor of GridStruct?
+        // GridStruct::setUp(input_, grids_);
+        GridStruct::addPointsToGrid(input_, grids_);
+
+        for (const auto &grid : grids_)
+        {
+            if (GridStruct::filterGrid(grid))
+                continue;
+
+            output.push_back(GridStruct::reduceGrid(grid));
+        }
+    }
+
+    PointCloud input_;
+    Grids grids_;
+    // ...
+};
+
+// ================== VoxelGrid.hpp ==================
+
+// structure of a single grid
+struct Voxel
+{
+    Eigen::Vector4f centroid;
+    float metric;
+    std::vector<size_t> point_indices;
+};
+
+// structure and operations of the grids
+
+using Voxels = std::unordered_map<size_t, Voxel>;
```

Comment:
Make this a template argument (but defaulted). In some cases, specially with integers, some other hash-maps are faster. This is solely for the the API of VoxelStructT

Reply:
Done by having `using`/`typedef` in `GridStruct`, I've tried other approach but didn't work